### PR TITLE
feat(product-assistant): a human in the loop for the taxonomy agent

### PR DESCRIPTION
**Pull request by skoob13**
## Problem

If the request is ambiguous or the taxonomy doesn't provide a clear picture, we should ask the user to clarify their request.

**Demo**

## Changes

- Add a checkpointer implementation on Django for LangGraph.
- Threads are now stored in the database (+ checkpoints, checkpoint writes, and checkpoint blobs).
- Implement a human in the loop for the taxonomy agent.
- Refactor how we restore a conversation since human messages can now appear in any order.
- Move the assistant to its endpoint.
- Use Pydantic for the state to simplify further state serialization.
- Add more tests.

Further work:
- Improve serialization of checkpoints. I've decided to split these PRs because this appears to be more complicated than I expected.
- Improve frontend UI, so users can easily search and add specific events to a prompt.
- Add eval tests.

## Does this work well for both Cloud and self-hosted?

No

## How did you test this code?

Manual testing, Storybook, more unit tests.

### CODE REVIEW

File: ee/models/assistant.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+from langgraph.checkpoint.serde.types import TASKS
+
+from posthog.models.team.team import Team
+from posthog.models.user import User
+from posthog.models.utils import UUIDModel
+
+
+class Conversation(UUIDModel):
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
+    team = models.ForeignKey(Team, on_delete=models.CASCADE)
```
Comment (skoob13) [inline ee/models/assistant.py:13]: Now, I think that there should be a ref to the project because property values can differ. @Twixes what do you think?

**Comment by skoob13** [comment]
The CI fails because there is a bug in migrations tests, which I've fixed in https://github.com/PostHog/posthog/pull/26928.

**Review by Twixes — COMMENTED** [review]
A mammoth PR so I might have missed something, but overall looks good, with a few comments!

File: frontend/src/types.ts [hunk 2]
```diff
@@ -4828,3 +4828,10 @@ export type ReplayTemplateVariableType = {
     filterGroup?: UniversalFiltersGroupValue
     noTouch?: boolean
 }
+
+/**
+ * Assistant Conversation
+ */
+export interface Conversation {
```
[the remark this answers is not in the document]
Comment (skoob13) [reply frontend/src/types.ts:4835]: Yes, the assistant now sends the conversation back in streaming if it's new. It is barebones now but will eventually grow.

File: ee/hogai/utils/types.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+class ReplaceMessages(list[AssistantMessageUnion]):
+    pass
+
+
+def add_messages(
+    left: Sequence[AssistantMessageUnion], right: Sequence[AssistantMessageUnion]
+) -> Sequence[AssistantMessageUnion]:
+    if isinstance(right, ReplaceMessages):
+        return list(right)
+    return list(left) + list(right)
```
[the remark this answers is not in the document]
Comment (skoob13) [reply ee/hogai/utils/types.py:31]: Removed both.

File: frontend/__snapshots__/scenes-app-max-ai--welcome-loading-suggestions--dark.png
[the remark this answers is not in the document]
Comment (skoob13) [reply frontend/__snapshots__/scenes-app-max-ai--welcome-loading-suggestions--dark.png:1]: Fixed. Replaced the mocked endpoint accidentally.

File: ee/hogai/django_checkpoint/checkpointer.py
Comment (Twixes) [inline ee/hogai/django_checkpoint/checkpointer.py:1]: Not important to shipping, but could be useful in terms of maintainability: Would it make sense to split this out into its own PyPi package, `langgraph_checkpoint_django`? (Could also be useful for others building on Django)

**Review by Twixes — APPROVED** [review]
Let's get this in

**Review by Twixes — APPROVED** [review]
Let's get this in

**Review by Twixes — COMMENTED** [review]
A mammoth PR so I might have missed something, but overall looks good, with a few comments!

**Comment by Twixes** [comment]
`dashboard.cy.ts` has failed 4 times, so something seems to be up 🤔 @skoob13

**Pull request merged**